codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -496,19 +520,49 @@ private void incrementPendingBytes(int numBytes) {
              */
             int write(int allowedBytes) {
                 int before = payload.size();
-                needFlush |= payload.write(Math.max(0, allowedBytes));
-                int writtenBytes = before - payload.size();
+                int writtenBytes = 0;
                 try {
-                    connectionState().incrementStreamWindow(-writtenBytes);
-                    incrementStreamWindow(-writtenBytes);
+                    boolean writeError = true;
+                    try {
+                        writing = true;
```

Comment:
If everything goes right we remove on frame at a time from the queue. If something goes wrong we call payload.error() which calls cancel() (with writing=false).

yes it's part of the documentation, but it's good practice to being able to handle the case. It doesn't cost us anything to do it. Also it's easier to reason about the correctness of the implementation, since we know that before and after a call to write the queue remains the same. This was @nmittler's idea btw.